libp2p/protocols/pubsub/pubsubpeer.nim

```diff
@@ -170,14 +172,24 @@ proc recvObservers*(p: PubSubPeer, msg: var RPCMsg) =
   if not (isNil(p.observers)) and p.observers[].len > 0:
     for obs in p.observers[]:
       if not (isNil(obs)): # TODO: should never be nil, but...
-        obs.onRecv(p, msg)
+        if not (isNil(obs.onRecv)):
+          obs.onRecv(p, msg)
+
+proc recvAndValidatedObservers*(p: PubSubPeer, msg: Message, msgId: MessageId) =
+  # trigger hooks
+  if not (isNil(p.observers)) and p.observers[].len > 0:
+    for obs in p.observers[]:
+      if not (isNil(obs)): # TODO: should never be nil, but...
```

Comment:
Which part? I added `if not (isNil(obs))` because it is run for all the other observers and if I understand correctly this case should be the same for all observers.

The `if not (isNil(obs.onValidated))` I added it here and for the other observers,  because if we register an observer and define only one of them (for example setting a proc for `onRecv` but leaving `onSend` empty), it would crash when trying to read from `nil` when attempting to run the undefined observer.

So added this to avoid this crash, and not have to define an empty proc for an observer that is not used.

Not sure if it makes sense

Reply:
I was talking about `if not (isNil(obs))`, there shouldn't be any `nil` inside this seq. Also, in the future please avoid using TODOs and create issues instead.

I'm also not sure about `observers*: ref seq[PubSubObserver] # ref as in smart_ptr`, it's the first time I see a ref being used like that in the codebase. @dryajov could you please advise here, Is it needed? I found this, but I'm not sure about it https://github.com/vacp2p/nim-libp2p/pull/156#discussion_r420042988

Reply:
`if not (isNil(p.observers))` shouldn't be necessary either.

Reply:
> I was talking about `if not (isNil(obs))`, there shouldn't be any `nil` inside this seq. Also, in the future please avoid using TODOs and create issues instead.

I just copy pasted how the observers are currently called (including the TODO comment) and added the `if not (isNil(obs.onValidated))` after I had crashes running tests when leaving one empty observer
https://github.com/vacp2p/nim-libp2p/blob/f8d4da6421b767727dd1501cb31b27636af87755/libp2p/protocols/pubsub/pubsubpeer.nim#L168-L180

Should I delete that check and the TODO from all 3 of them?

Reply:
You can leave the others for now, but remove them from yours. Thanks.

Reply:
Done, I did have to leave the
```
if not (isNil(obs.onValidated))
```

because otherwise I did get a
```
SIGSEGV: Illegal storage access. (Attempt to read from nil?)
```
on a test in which other observers are registered but that one was left empty

Reply:
`if not (isNil(obs.onValidated))` is fine, although it could be an Option. But it's not necessary to change it now. 

I guess the `ref` purpose in `ref seq[PubSubObserver]` is to set `observers` in `PubSub` and have a reference to it in the `PubSubPeer` instances.